timemachine/fe/interaction_group_traj.py

```diff
@@ -0,0 +1,187 @@
+
+    return vmap(within_cutoff)(x_env)
+
+
+class InteractionGroupTraj:
+    def __init__(self, xs: Array, box_diags: Array, ligand_idxs: Array, env_idxs: Array, cutoff=1.2, verbose=True):
+        r"""support [U_ig(x; params) for x in traj]
+
+        where U_ig = \sum_i \sum_j pair_fxn(||x_j - x_j||; params_i, params_j)
+
+        (with i summing over ligand_idxs, j summing over env_idxs, and pair_fxn(r) == 0 when r >= cutoff)
+
+        Parameters
+        ----------
+        xs : array of particle coordinates [T, N, 3]
+        box_diags : positive float array of box diagonals [T, 3]
+        ligand_idxs: int array (subset of arange(N))
+        env_idxs : int array (subset of arange(N))
+        cutoff : float
+        verbose: bool
+
+        Notes
+        -----
+        * Constructor precomputes neighborlist using a naive jax implementation, which is expensive, but done once
+            TODO: extract from GPU neighborlist
+        * Assumes pair_fxn(r) == 0 when r >= cutoff, but does not confirm or enforce this
+        """
+        self.cutoff = cutoff
+        self.n_frames = len(xs)
+        self.ligand_idxs = ligand_idxs
+        all_env_idxs = env_idxs
+        num_lig, num_env = len(ligand_idxs), len(env_idxs)
+
+        # TODO[symmetry]:
+        #   maybe generalize so we select a subset of ligand atoms too...
+        #   (currently: assume we always look at all ligand atoms)
+        self.xs_lig = xs[:, ligand_idxs]
+        _xs_env = xs[:, env_idxs]  # will select subsets...
+
+        if verbose:
+            print(f"precomputing neighborlist on ({num_lig}, {num_env}) interaction group, at cutoff={cutoff}")
+
+        # note: vmap here can consume excessive memory if len(env_idxs) * len(xs) is large -> python loop
+        # mask = vmap(env_mask_within_cutoff, (0,0,0,None))(_xs_env, self.xs_lig, boxes, cutoff)
+        def f(x_env, x_lig, box):
+            return env_mask_within_cutoff(x_env, x_lig, box, cutoff)
+
+        mask = np.array([f(_xs_env[i], self.xs_lig[i], np.diag(box_diags[i])) for i in range(self.n_frames)])
+
+        padded_num_env_atoms = mask.sum(1).max()
+
+        if verbose:
+            num_stored = padded_num_env_atoms + len(ligand_idxs)
+            max_nbrs, mean_nbrs = padded_num_env_atoms, mask.sum(1).mean()
+            msg = f"""
+                saving {(xs.shape[1] / num_stored):.2f}x on storage
+                (relative to storing all env atoms)
+                padding to max_nbrs = {max_nbrs}
+                (~{max_nbrs/mean_nbrs:.2f}x larger than unpadded neighbor list)
+            """
+            print(msg)
+
+        idxs_within_env_block = np.argsort(mask, axis=1)[:, -padded_num_env_atoms:]
+        self.selected_env_idxs = jnp.array(all_env_idxs[idxs_within_env_block], dtype=jnp.uint32)
+
+        self.xs_env = np.array([_x_env[idxs] for (_x_env, idxs) in zip(_xs_env, idxs_within_env_block)])
+        self.box_diags = box_diags
+
+    def to_npz(self, fname):
```

Comment:
to/from a dictionary would be more flexible

Reply:
Good idea : added in https://github.com/proteneer/timemachine/pull/1383/commits/c19938c3db299a4b3feee7e67131f45cf51660fa

(to_npz / from_npz become short functions in terms of to_dict / from_dict)